File: src/AutoNate.Spa/src/components/aql-editor/aqlLexer.ts
// TypeScript port of AqlLexer.cs. Kept 1:1 with the C# source so the
// SPA's autocomplete state machine sees the same tokens the backend
// parser will see. Tokens carry their absolute offsets so callers can
// map back to the editor's caret.
//
// Unlike the C# lexer, this one never throws — unterminated strings
// and stray characters are emitted as best-effort tokens so the editor
// can still offer completions mid-edit (when the buffer is necessarily
// malformed).

export type TokenKind =
  | "keyword"
  | "identifier"
  | "string"
  | "number"
  | "relativeDate"
  | "bool"
  | "null"
  | "operator"
  | "lparen"
  | "rparen"
  | "comma"
  | "unknown"
  | "eof";

export type Token = {
  kind: TokenKind;
  // The canonical lexeme: keywords/bools/null are upper-cased; strings
  // are the un-quoted, escape-processed value; numbers and relative
  // dates are the raw source text; identifiers are the raw source text.
  lexeme: string;
  // Source range. `start` is the index of the first character of this
  // token in the input; `end` is one past the last character (so
  // `source.slice(start, end)` round-trips the original text).
  start: number;
  end: number;
};

const KEYWORDS = new Set([
  "FROM", "WHERE", "ORDER", "BY", "ASC", "DESC",
  "COLUMNS", "GROUP", "LIMIT",
  "AND", "OR", "AS"
]);

const REL_DATE_UNITS = new Set(["h", "d", "w", "m", "y", "H", "D", "W", "M", "Y"]);

export function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  while (i < source.length) {
    const c = source[i];

    if (isWhitespace(c)) {
      i++;
      continue;
    }

    // Sign-prefixed number/relative-date. `-Foo` is never a number;
    // `-2w` always is a relative date.
    if ((c === "-" || c === "+") && i + 1 < source.length && isDigit(source[i + 1])) {
      tokens.push(readNumericOrRelativeDate(source, i));
      i = tokens[tokens.length - 1].end;
      continue;
    }
    if (isDigit(c)) {
      tokens.push(readNumericOrRelativeDate(source, i));
      i = tokens[tokens.length - 1].end;
      continue;
    }

    if (c === '"' || c === "'") {
      tokens.push(readString(source, i));
      i = tokens[tokens.length - 1].end;
      continue;
    }

    if (c === "(") { tokens.push({ kind: "lparen", lexeme: "(", start: i, end: i + 1 }); i++; continue; }
    if (c === ")") { tokens.push({ kind: "rparen", lexeme: ")", start: i, end: i + 1 }); i++; continue; }
    if (c === ",") { tokens.push({ kind: "comma", lexeme: ",", start: i, end: i + 1 }); i++; continue; }

    if (c === "=" || c === "<" || c === ">" || c === "!" || c === "~") {
      tokens.push(readOperator(source, i));
      i = tokens[tokens.length - 1].end;
      continue;
    }

    if (isIdentStart(c)) {
      tokens.push(readIdentifierOrKeyword(source, i));
      i = tokens[tokens.length - 1].end;
      continue;
    }

    tokens.push({ kind: "unknown", lexeme: c, start: i, end: i + 1 });
    i++;
  }
  tokens.push({ kind: "eof", lexeme: "", start: source.length, end: source.length });
  return tokens;
}

// ---- Helpers ---------------------------------------------------------

function readNumericOrRelativeDate(source: string, start: number): Token {
  let i = start;
  if (source[i] === "-" || source[i] === "+") i++;
  while (i < source.length && isDigit(source[i])) i++;

  let hasFraction = false;
  if (i < source.length && source[i] === "." && i + 1 < source.length && isDigit(source[i + 1])) {
    hasFraction = true;
    i++;
    while (i < source.length && isDigit(source[i])) i++;
  }

  if (!hasFraction && i < source.length) {
    const ch = source[i];
    // Relative-date suffix must NOT be followed by another ident char,
    // so `2hours` stays as Number(2) + Identifier(hours).
    if (REL_DATE_UNITS.has(ch) && (i + 1 >= source.length || !isIdentPart(source[i + 1]))) {
      i++;
      return { kind: "relativeDate", lexeme: source.slice(start, i), start, end: i };
    }
  }

  return { kind: "number", lexeme: source.slice(start, i), start, end: i };
}

function readString(source: string, start: number): Token {
  const quote = source[start];
  let i = start + 1;
  let out = "";
  while (i < source.length && source[i] !== quote) {
    if (source[i] === "\\" && i + 1 < source.length) {
      const next = source[i + 1];
      out += escapeMap[next] ?? next;
      i += 2;
      continue;
    }
    out += source[i];
    i++;
  }
  // Tolerate unterminated strings (mid-typing). The token ends at
  // wherever we ran out of input; callers treat "we're inside an open
  // string" specially.
  const end = i < source.length ? i + 1 : i;
  return { kind: "string", lexeme: out, start, end };
}

const escapeMap: Record<string, string> = {
  n: "\n",
  t: "\t",
  r: "\r",
  "\\": "\\",
  "\"": "\"",
  "'": "'"
};

function readOperator(source: string, start: number): Token {
  const c = source[start];
  if (c === "!" && start + 1 < source.length && source[start + 1] === "=") {
    return { kind: "operator", lexeme: "!=", start, end: start + 2 };
  }
  if ((c === "<" || c === ">") && start + 1 < source.length && source[start + 1] === "=") {
    return { kind: "operator", lexeme: source.slice(start, start + 2), start, end: start + 2 };
  }
  if (c === "!") {
    // Stand-alone `!` is invalid AQL — emit as unknown so the detector
    // doesn't mistake it for an operator the backend will accept.
    return { kind: "unknown", lexeme: "!", start, end: start + 1 };
  }
  return { kind: "operator", lexeme: c, start, end: start + 1 };
}

function readIdentifierOrKeyword(source: string, start: number): Token {
  let i = start;
  while (i < source.length && isIdentPart(source[i])) i++;
  const lexeme = source.slice(start, i);
  const upper = lexeme.toUpperCase();
  if (upper === "TRUE" || upper === "FALSE") {
    return { kind: "bool", lexeme: upper, start, end: i };
  }
  if (upper === "NULL") {
    return { kind: "null", lexeme: "NULL", start, end: i };
  }
  if (KEYWORDS.has(upper)) {
    return { kind: "keyword", lexeme: upper, start, end: i };
  }
  return { kind: "identifier", lexeme, start, end: i };
}

function isWhitespace(c: string): boolean {
  return c === " " || c === "\t" || c === "\n" || c === "\r" || c === "\f" || c === "\v";
}

function isDigit(c: string): boolean {
  return c >= "0" && c <= "9";
}

function isIdentStart(c: string): boolean {
  return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z") || c === "_";
}

function isIdentPart(c: string): boolean {
  return isIdentStart(c) || isDigit(c);
}

File: src/AutoNate.Web/Services/Query/AqlLexer.cs
using System.Text;

namespace AutoNate.Web.Services.Query;

internal static class AqlLexer
{
    // Reserved words. Case-insensitive at the source; the lexer upper-cases
    // the lexeme before classifying so the parser only needs to compare to
    // the canonical form.
    private static readonly HashSet<string> Keywords = new(StringComparer.Ordinal)
    {
        "FROM", "WHERE", "ORDER", "BY", "ASC", "DESC",
        "COLUMNS", "GROUP", "LIMIT",
        "AND", "OR", "AS"
    };

    public static IReadOnlyList<AqlToken> Tokenize(string source)
    {
        ArgumentNullException.ThrowIfNull(source);

        var tokens = new List<AqlToken>();
        var i = 0;
        while (i < source.Length)
        {
            var c = source[i];

            if (char.IsWhiteSpace(c)) { i++; continue; }

            // Numbers and relative dates. A sign prefix counts only when
            // immediately followed by a digit. `-Foo` is never a number; `-2w`
            // always is a RelativeDate.
            if ((c == '-' || c == '+') && i + 1 < source.Length && char.IsDigit(source[i + 1]))
            {
                tokens.Add(ReadNumericOrRelativeDate(source, ref i));
                continue;
            }
            if (char.IsDigit(c))
            {
                tokens.Add(ReadNumericOrRelativeDate(source, ref i));
                continue;
            }

            if (c == '"' || c == '\'')
            {
                tokens.Add(ReadString(source, ref i));
                continue;
            }

            if (c == '(') { tokens.Add(new AqlToken(TokenKind.LParen, "(", i)); i++; continue; }
            if (c == ')') { tokens.Add(new AqlToken(TokenKind.RParen, ")", i)); i++; continue; }
            if (c == ',') { tokens.Add(new AqlToken(TokenKind.Comma, ",", i)); i++; continue; }

            if (c == '=' || c == '<' || c == '>' || c == '!' || c == '~')
            {
                tokens.Add(ReadOperator(source, ref i));
                continue;
            }

            if (IsIdentStart(c))
            {
                tokens.Add(ReadIdentifierOrKeyword(source, ref i));
                continue;
            }

            throw new AqlValidationException($"Unexpected character '{c}' at position {i}.");
        }

        tokens.Add(new AqlToken(TokenKind.Eof, string.Empty, source.Length));
        return tokens;
    }

    private static AqlToken ReadNumericOrRelativeDate(string source, ref int i)
    {
        var start = i;
        if (source[i] == '-' || source[i] == '+') i++;
        while (i < source.Length && char.IsDigit(source[i])) i++;

        // Optional fractional part.
        var hasFraction = false;
        if (i < source.Length && source[i] == '.' && i + 1 < source.Length && char.IsDigit(source[i + 1]))
        {
            hasFraction = true;
            i++;
            while (i < source.Length && char.IsDigit(source[i])) i++;
        }

        // Relative-date suffix: h | d | w | m | y, not followed by another
        // identifier char (so `2hours` stays a Number(2) + Identifier(hours)).
        if (!hasFraction && i < source.Length)
        {
            var ch = source[i];
            if ((ch == 'h' || ch == 'd' || ch == 'w' || ch == 'm' || ch == 'y' ||
                 ch == 'H' || ch == 'D' || ch == 'W' || ch == 'M' || ch == 'Y')
                && (i + 1 >= source.Length || !IsIdentPart(source[i + 1])))
            {
                i++;
                return new AqlToken(TokenKind.RelativeDate,
                    source.Substring(start, i - start), start);
            }
        }

        return new AqlToken(TokenKind.Number,
            source.Substring(start, i - start), start);
    }

    private static AqlToken ReadString(string source, ref int i)
    {
        var quote = source[i];
        var start = i;
        i++;
        var sb = new StringBuilder();
        while (i < source.Length && source[i] != quote)
        {
            if (source[i] == '\\' && i + 1 < source.Length)
            {
                var next = source[i + 1];
                sb.Append(next switch
                {
                    'n' => '\n',
                    't' => '\t',
                    'r' => '\r',
                    '\\' => '\\',
                    '"' => '"',
                    '\'' => '\'',
                    _ => next
                });
                i += 2;
                continue;
            }
            sb.Append(source[i]);
            i++;
        }
        if (i >= source.Length)
        {
            throw new AqlValidationException(
                $"Unterminated string literal starting at position {start}.");
        }
        i++; // skip closing quote
        return new AqlToken(TokenKind.String, sb.ToString(), start);
    }

    private static AqlToken ReadOperator(string source, ref int i)
    {
        var start = i;
        var c = source[i];
        if (c == '!' && i + 1 < source.Length && source[i + 1] == '=')
        {
            i += 2;
            return new AqlToken(TokenKind.Operator, "!=", start);
        }
        if ((c == '<' || c == '>') && i + 1 < source.Length && source[i + 1] == '=')
        {
            var op = source.Substring(i, 2);
            i += 2;
            return new AqlToken(TokenKind.Operator, op, start);
        }
        if (c == '!')
        {
            throw new AqlValidationException(
                $"Unexpected '!' at position {start}. Did you mean '!='?");
        }
        i++;
        return new AqlToken(TokenKind.Operator, c.ToString(), start);
    }

    private static AqlToken ReadIdentifierOrKeyword(string source, ref int i)
    {
        var start = i;
        while (i < source.Length && IsIdentPart(source[i])) i++;
        var lexeme = source.Substring(start, i - start);
        var upper = lexeme.ToUpperInvariant();

        if (upper == "TRUE" || upper == "FALSE")
        {
            return new AqlToken(TokenKind.Bool, upper, start);
        }
        if (upper == "NULL")
        {
            return new AqlToken(TokenKind.Null, "NULL", start);
        }
        // NOW is desugared at parse time into AqlRelativeDate(0,'d') so every
        // entity's existing relative-date plumbing handles it for free. The
        // lexeme stays "NOW" so error messages quote what the user wrote.
        if (upper == "NOW")
        {
            return new AqlToken(TokenKind.RelativeDate, "NOW", start);
        }
        if (Keywords.Contains(upper))
        {
            return new AqlToken(TokenKind.Keyword, upper, start);
        }
        return new AqlToken(TokenKind.Identifier, lexeme, start);
    }

    private static bool IsIdentStart(char c) =>
        char.IsLetter(c) || c == '_';

    private static bool IsIdentPart(char c) =>
        char.IsLetterOrDigit(c) || c == '_';
}
